users/stdiscosrv.rst

```diff
@@ -205,20 +205,27 @@ allows:
 - Use of a subdomain name without requiring a port number added to the URL
 - Sharing an SSL certificate with multiple services on the same server
 
+Note that after this configuration, **clients have to omit the** :code:`?id=...`
+**parameter from the discovery server URL on their configuration**. Client-side
+validation will be done by checking the visible proxy server's HTTPS certificate.
```

Comment:
I don't think this is true, the ID is used for verifying discovery server is who it says it is.
It has to be a valid TLS cert, or if it's self signed, the hash of the cert is turned into an ID.

Reply:
That's odd.
When I use this configuration, I get the following when I define `?id=<discovery server ID>`:

```[THISDEV] 2020/05/12 22:10:05.990327 global.go:155: DEBUG: globalClient.Lookup https://<myserver>.org/syncthingdiscosrv?device=SOMEDEV security: incorrect device id
[THISDEV] 2020/05/12 22:10:05.990827 cache.go:124: DEBUG: lookup results for SOMEDEV
[THISDEV] 2020/05/12 22:10:05.991038 cache.go:125: DEBUG:   addresses:  []
[THISDEV] 2020/05/12 22:10:06.019513 global.go:238: DEBUG: announce POST: security: incorrect device id
```

When I do not set such ID:
```[THISDEV] 2020/05/12 21:12:50.765689 cache.go:102: DEBUG: lookup for SOMEDEV at global@https://<myserver>.org/syncthingdiscosrv
[THISDEV] 2020/05/12 21:12:50.765739 cache.go:103: DEBUG:   addresses: [relay://<myserver>:22067/?id=RELAYID]
[THISDEV] 2020/05/12 21:12:50.765770 cache.go:124: DEBUG: lookup results for SOMEDEV
[THISDEV] 2020/05/12 21:12:50.765788 cache.go:125: DEBUG:   addresses:  [relay://<myserver>.org:22067/?id=RELAYID]
[THISDEV] 2020/05/12 21:12:50.773228 global.go:243: DEBUG: announce POST: 204 No Content
[THISDEV] 2020/05/12 21:12:50.773275 global.go:270: DEBUG: announce Reannounce-After: 3392 <nil>
```
Note especially the POST parts.

Using tcpdump, I see the discovery server replying with HTTP 204 for both cases.
So, this is the client not being able to validate what comes from the server when the id is set. The server is indeed considering all requests valid.

By reading the page in https://docs.syncthing.net/specs/globaldisco-v3.html#authentication , I see the following:
> The “id” parameter is not, in fact, sent to the discovery server - it’s used by Syncthing itself to know which certificate to expect on the server side.

So, it seems that the protocol is something like:
1. Client sends POST to the discovery server
2. Discovery server replies with a 204.
3. Because that response comes through HTTPS, the client will try to validate the certificate associated with that response against the locally-configured ID.

This means that, when the connection is coming from a proxy (that by definition creates a new HTTPS connection), the proxy's certificate will never be possible to validate against the client's locally-configured ID for the discovery server.
This will be true for any proxy server, be it Apache, Nginx or any other.

The only validation done on the client side, when no ID is configured, appears to be that the certificate associated to the proxy's HTTPS certificate is valid. I checked this briefly by setting an invalid certificate on the server. The syncthing client got this:
```
[THISDEV] 2020/05/12 22:48:25.412384 global.go:238: DEBUG: announce POST: Post https://<myserver>.org/syncthingdiscosrv: x509: certificate has expired or is not yet valid
```
Additionally, and as you say:
> Discovery server has to have a valid TLS cert, or if it's self signed, the hash of the cert is turned into an ID, for client to verify there is no man in the middle.

With a proxy setup, we want to make sure there is a man in the middle. The assurance the assurance can get is if the client validates the configured "man-in-the-middle"'s (our proxy server) HTTPS certificate is valid regarding its DNS name. And the client validating this certificate is what gives us the same confidence level we get when the ID is used in the no-proxy setup.

Is this correct?
If you can confirm this assessment, I can try to make the doc simpler to explain this.

Reply:
It is correct, but removing the ID is not universally the answer.
It's only the answer if the proxy serves a valid certificate, if it's not a valid certificate, then the user should somehow compute the id of the cert and pin the cert that way.

Reply:
I rephrased this paragraph. What do you think?